### ISSUE TITLE
al_save_bitmap with screen is wrong

### ISSUE THREAD

**Issue by katastic** [liballeg/allegro5#1337]
I just posted on A.CC but the server is currently half-dead so here we go.

````D
al_save_bitmap("screen.png", al_get_backbuffer(al_display));	
````

Saving a screenshot of the screen is wrong (at least on Linux).  I just realized **it's dumping a alpha channel**! Which, AFAIK, is meaningless for a screen.

gnome-screenshot yields this:

![Screenshot from 2022-05-23 03-40-28](https://user-images.githubusercontent.com/11449593/169782148-37bab798-30fe-4e45-bebe-866a266e140b.png)

al_save_bitmap loading with GIMP does this: (note alpha channel visible)
![image](https://user-images.githubusercontent.com/11449593/169782319-981bad04-c462-4c4f-9194-8880c0368d2f.png)

**Comment by katastic** [comment]
Saving to a TARGA is also affected since TARGA supports alpha channels. PCX and BMP output fine.

Also, DDS doesn't appear to output any file at all even though the documentation says PCX, BMP, TGA and DDS are supported always.

**Comment by SiegeLord** [comment]
What does `al_get_pixel` return on one of the black regions of your screen?

**Comment by katastic** [comment]
````D
RGBA
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
0.000000 0.000000 0.000000 0.623529
````

**Comment by katastic** [comment]
Not sure if anyone else needs this, but I wrote this D function to work around the problem.
````D
void al_save_screen(string path)
	{
	auto disp = al_get_backbuffer(al_display);
	auto w = al_get_bitmap_width(disp);
	auto h = al_get_bitmap_height(disp);
	ALLEGRO_BITMAP* temp = al_create_bitmap(w, h);
	al_lock_bitmap(temp, al_get_bitmap_format(temp), ALLEGRO_LOCK_WRITEONLY);
	al_lock_bitmap(disp, al_get_bitmap_format(temp), ALLEGRO_LOCK_READONLY); // makes HUGE difference (6.4 seconds vs 270 milliseconds)
	al_set_target_bitmap(temp);
	for(int j = 0; j < h; j++)
		for(int i = 0; i < w; i++)
			{
			auto pixel = al_get_pixel(disp, i, j);
			pixel.a = 0.0; // remove alpha
			al_put_pixel(i, j, pixel);
			}
	al_unlock_bitmap(disp);
	al_unlock_bitmap(temp);
	al_save_bitmap(path.toStringz, temp);
	al_reset_target();
	al_destroy_bitmap(temp);
	}
````

Which runs in ~250 ms on my netbook vs ~170 ms for al_save_bitmap(). It would be very few changes to port it to C.

**Comment by pedro-w** [comment]
I don't understand why compositing it onto an opaque black bitmap doesn't work. I (independently) wrote this
```
void save_opaque(const char* fi, ALLEGRO_BITMAP* b) {
  // Create clone bitmap
  int w = al_get_bitmap_width(b);
  int h = al_get_bitmap_height(b);
  ALLEGRO_BITMAP* tmp = al_create_bitmap(w, h);
  // Remember current target
  ALLEGRO_BITMAP* old = al_get_target_bitmap();
  al_set_target_bitmap(tmp);
  // Clear to opaque black
  al_clear_to_color(al_map_rgba_f(0.0f, 0.0f, 0.0f, 1.0f));
  al_draw_bitmap(b, 0.0f, 0.0f, 0);
  // Save and clean up
  al_save_bitmap(fi, tmp);
  al_destroy_bitmap(tmp);
  // Restore target
  al_set_target_bitmap(old);
}
```

to be called like `save_opaque("shot.png", al_get_backbuffer(al_get_current_display()));`
and it 'worked' - but on my system the backbuffer never had an alpha channel anyway.

Also is it possible to create a colour with `ALLEGRO_COLOR(0,0,0,1)` rather than `al_map_rgba_f`, is that a D thing?

**Comment by SiegeLord** [comment]
That's a D-syntax, but in C/C++ you can use `{0, 0, 0, 1}` or w/e other struct initializer syntax. `ALLEGRO_COLOR` has all of its fields public.

As for this issue, I'd like to know why the screen has alpha < 1. Maybe I need to try to compile your game and see what's up.

**Comment by katastic** [comment]
If there's a snippet of code you want me to try on my machine, I can do that. I'm going to re-try the provided clear_to_color() code tonight to double check I'm not mixing up screenshots with code. 

And maybe do some al_get_pixel logging between each draw layer of the "space" textures.

The images I'm drawing (of which you can grab from my dgravity repo above), all have transparency layers themselves, in addition to drawing them with al_draw_tinted_bitmap() alpha values. 

https://github.com/katastic/dgravity/blob/5c469603180fc4304c5d28b251c2553c584a85ef/data/seamless_space.png?raw=true

Also the blending mode used (perhaps this is key):
````D
al_set_blender(ALLEGRO_BLEND_OPERATIONS.ALLEGRO_ADD, ALLEGRO_ALPHA, ALLEGRO_INVERSE_ALPHA);
````

**Comment by pedro-w** [comment]
Right. I was able to compile and run your game on MacOS and get the same issue with the translucentn screenshot, so it's not a platform dependent thing. (D compiles amazingly quick, it's really impressive!)
In a nutshell I'd expect drawing a translucent image over an opaque background to give me an opaque bitmap with the image blended in. In my mini-program above, drawing the backbuffer onto a background just replaces the background with the screen image, alpha and all.
